0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+
+For example, a circuit with unbound parameters (or in OpenQASM3 terms, a circuit with `input`s) alone could never qualify as a task for any primitive because there is not enough information to execute it, namely, numeric parameter binding values. On the other hand, conceptually, a circuit with no unbound parameters (i.e. an circuit with no `input`s) alone could form a task for a hypothetical primitive that just runs circuits and returns counts. This suggests a natural base for all tasks:
+
+```python
+@dataclass
+class BaseTask:
+    circuit: QuantumCircuit
+```
+
+This base class can be added to Qiskit if and when needed as a non-breaking change, and has been included here mainly to make the concepts clear, rather than to indicate that we need such a base class immediately.
+Most relevant, for the `Estimator` primitive, in order to satisfy the definition as stated above, we propose the task structure
+
+```python
+class EstimatorTask(BaseTask):
+    observables: ObservablesArray = None
+    parameter_values: BindingsArray = None, 
+```
+
+We expect the formal primitive API and primitive implementations to have a strong sense of tasks, but we will not demand that users construct them manually in Python as they are little more than simple data containers (incidentically implemented as dataclasses above), and we do not wish to overburden them with types. This is discussed further in [Type Coercion Strategy](#type-coercion-strategy).
+
+### BindingsArray <a name="bindingsarray"></a>
+
+It is common for a user to want to do a sweep over parameter values, that is, to execute the same parametric circuit with many different parameter binding sets. 
+We propose a new class `BindingsArray` that stores multiple sets of parameters, which can be bound to a circuit. 
+For example, if a circuit has 200 parameters, and a user wishes to execute the circuit for 50 different sets of values, then a single instance of `BindingsArray` could represent 50 sets of 200 parameter values. 
+Moreover, we propose that it be array-like, so that, in this example, the `BindingsArray` instance would be one-dimensional and have shape equal `(50,)`. 
+
+The proposed `BindingsArray` object would support, at a minimum, the following constructor examples for a circuit with three input parameters `a`, `b`, and `c`, where we will use `<>` to denote some `array_like` of the specified shape:
+
+```python
+# a single value for each a, b, and c; 0-dimensional, and compatible with current interface
+BindingsArray([0.0, 1.0, 2.0])
+
+# a single value for each a, b, and c; 0-dimensional, and compatible with current interface
+# satisfies user wish to be able to specify values by name
+BindingsArray(kwvals={Parameter("a"): 0.0, Parameter("b"): 1.0, Parameter("c"): 2.0})
```

Comment:
This line uses `kwvals`, but a below line uses `kwargs`.  Is there a difference?

At first I wanted to replace `kwvals=` with the standard python syntax for passing kwargs (`**`), but then I realized that a `Parameter` is not a valid kwarg; it has to be a string.

I do prefer a name other than kwargs, which already has a precise meaning in python.  Perhaps kwvals is better.  (It's still a non-standard abbreviation, though.)

Reply:
> This line uses kwvals, but a below line uses kwargs. Is there a difference?

`kwargs` is the typo, let me fix that while we continue to discuss.

> At first I wanted to replace kwvals= with the standard python syntax for passing kwargs (**), but then I realized that a Parameter is not a valid kwarg; it has to be a string.

There are two things that make it not a string: the ability to do things like `{(Parameter("a"), Parameter("b")): <..., 2>}`, and as you say, `Parameter` not being a string for those cases where you include an array for each parameter separately. I think the former limitation is the stronger argument for avoiding kwargs in the signature (as opposed to a dictionary argument).

I meant to bring up, so thanks for reminding me (and maybe @jakelishman would have an opinion): should these be strings or `Parameters`? This is more from a "where is qiskit going with parameters" perspective than from the issue you raised, though the issue you raised is relevant.

> Perhaps kwvals is better.

Right now it's `vals` and `kwvals`. Happy to consider alternatives.

Reply:
Okay, I think pure strings is probably a bit cleaner too, I'll make the changes in that direction.

I don't want to lose the ability `{("a", "b"): <..., 2>}`, and my personal preference is to not have too many ways to do the same thing, so I'm in favor of not introducing direct unpacked at this time, with the possibility of introducing it as an enhancement in the future.